File: 0020-valid-parentheses/stack/python/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        bracket_mapping = { "]": "[", ")": "(", "}": "{"}  # close -> open
        stack = []

        for ele in s:
            if ele in bracket_mapping:      # close                        
                if stack and stack[-1] == bracket_mapping[ele]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(ele)       # open

        return not stack

File: 0020-valid-parentheses/stack/java/solution.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> bracketMap = new HashMap<>(); // close -> open
        bracketMap.put(')', '(');
        bracketMap.put('}', '{');
        bracketMap.put(']', '[');

        Stack<Character> openBracketStack = new Stack<>();
        
        for (char ele : s.toCharArray()) {
            if (bracketMap.containsKey(ele)) {  // close
                if (openBracketStack.isEmpty() || openBracketStack.pop() != bracketMap.get(ele)) {
                    return false;
                }
            }
            else {      // open
                openBracketStack.push(ele);
            }
        }
        return openBracketStack.isEmpty();
    }
}
